### PR TITLE
api-getway/fix/handle_errors_in_enhanced_way

### PR DESCRIPTION
**Pull request by Mohammed-Alhams**

### CODE REVIEW

**Review by Youssef-Ramadan-Y — APPROVED** [review]

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/usecase/LoginUserUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    override suspend fun getAccessToken(): String {
+        return localGateway.getAccessToken()
+    }
+
+    override suspend fun getRefreshToken(): String {
+        return localGateway.getRefreshToken()
+    }
+
+    override suspend fun loginUser(username: String, password: String, keepLoggedIn: Boolean): UserTokens {
+        return remoteGateway.loginUser(username, password, keepLoggedIn)
```
Comment (AzizaHelmy) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/usecase/LoginUserUseCase.kt:43]: Why are you passing the local flag to login method?!!

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/usecase/LoginUserUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package org.thechance.common.domain.usecase
+
+import org.thechance.common.domain.entity.UserTokens
+import org.thechance.common.domain.getway.ILocalGateway
+import org.thechance.common.domain.getway.IRemoteGateway
+
+interface ILoginUserUseCase {
```
Comment (AzizaHelmy) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/usecase/LoginUserUseCase.kt:7]: I think we need to add the method for store the local flag ( keep me logged in) also

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/getway/ILocalGateway.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package org.thechance.common.domain.getway
+
+
+interface ILocalGateway {
```
Comment (AzizaHelmy) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/getway/ILocalGateway.kt:4]: Why local gateway???
I understand that it's like the repo,then one place contains the difference sources right?!! Or it's like the data source so we should to separate it? "((
I'm confused"((

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/entity/UserTokens.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package org.thechance.common.domain.entity
+
+data class UserTokens(
```
Comment (AzizaHelmy) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/entity/UserTokens.kt:3]: Why did you separate this class? What happens if we add this attributes to the user class (login)
?!!

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/local/local_dto/KeepUserLoggedInLocalDto.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package org.thechance.common.data.local.local_dto
+
+class KeepUserLoggedInLocalDto(
```
Comment (AzizaHelmy) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/local/local_dto/KeepUserLoggedInLocalDto.kt:3]: Why?!!
I think it should be in the login

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/di/UseCaseModule.kt [hunk 6]
```diff
@@ -21,4 +12,5 @@ val UseCaseModule = module {
     singleOf(::CreateNewTaxiUseCase) { bind<ICreateNewTaxiUseCase>() }
     singleOf(::FindTaxiByUserNameUseCase) { bind<IFindTaxiByUsernameUseCase>() }
     singleOf(::GetRestaurantsUseCase) { bind<IGetRestaurantsUseCase>() }
-}
+    singleOf(::LoginUserUseCase) { bind<ILoginUserUseCase>() }
```
Comment (AzizaHelmy) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/di/UseCaseModule.kt:15]: what is the defferent between that and this way?
 single<ILoginUserUseCase> { LoginUserUseCase(get()) }

**Review by AzizaHelmy — CHANGES_REQUESTED** [review]
need to discus first befor approve merging for dashboard work

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/usecase/LoginUserUseCase.kt (hunk 1 above)
Comment (Mohammed-Alhams) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/domain/usecase/LoginUserUseCase.kt:43]: It's the usecase, where we can access the local and remote gateways, and It's a rule of our app to save the login preferences

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/di/UseCaseModule.kt (hunk 6 above)
Comment (Mohammed-Alhams) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/di/UseCaseModule.kt:15]: Bareq suggested that

Comment (Mohammed-Alhams) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/di/UseCaseModule.kt:15]: cause its more clear

**Review by HaidyAbuGom3a — APPROVED** [review]

**Review by ay0o5h — APPROVED** [review]

**Review by AzizaHelmy — APPROVED** [review]
Great effort, Masha'a Allah 👏

**Pull request merged**